src/lem/constrainer.rs

```diff
@@ -44,14 +44,70 @@ impl<F: LurkField> AllocationManager<F> {
             }
         }
     }
+
+    /// Calls `get_or_alloc_num` to allocate tag and hash for a pointer.
+    pub(crate) fn get_or_alloc_ptr<CS: ConstraintSystem<F>>(
+        &mut self,
+        cs: &mut CS,
+        z_ptr: &ZPtr<F>,
+    ) -> Result<AllocatedPtr<F>> {
+        Ok(AllocatedPtr::from_parts(
+            &self.get_or_alloc_num(cs, z_ptr.tag.to_field())?,
+            &self.get_or_alloc_num(cs, z_ptr.hash)?,
+        ))
+    }
+}
+
+#[derive(Default)]
+struct SlotData {
+    max_slots: usize,
+    constraints_data: Vec<(Boolean, usize, MetaPtr, Tag)>,
+}
+
+#[derive(Default)]
+struct HashSlots<F: LurkField> {
+    hash2_alloc: Vec<(usize, AllocatedPtr<F>, AllocatedPtr<F>)>,
+    hash2_data: SlotData,
+    hash3_alloc: Vec<(usize, AllocatedPtr<F>, AllocatedPtr<F>, AllocatedPtr<F>)>,
+    hash3_data: SlotData,
+    hash4_alloc: Vec<(
+        usize,
+        AllocatedPtr<F>,
+        AllocatedPtr<F>,
+        AllocatedPtr<F>,
+        AllocatedPtr<F>,
+    )>,
+    hash4_data: SlotData,
+}
```

Comment:
What if this whole thing became
```
struct SlotAllocation {
   idx: usize, // do we even need this?
  preimage: Vec<AllocatedPtr<F>>
}

struct HashSlots<F: LurkField> {
   arity_allocations: Vec<SlotAllocations<F>>
}
```
That would make all the code below much simpler, since the cases could be treated uniformly.

If you want to preserve more type information, you could add an enum with size information that holds typed instances of `SlotAllocation`, like
```
enum AllowedAllocations<F: LurkField> {
   A2(SlotAllocation<F, 2>,
   A3(SlotAllocation<F, 3>,
   …
}
```
Or something. Just musing. Boiling this all down to the essence would be nice, since I think the underlying logic is relatively minimal but a bit picky. Just seeing it written out clearly would simplify understanding, auditing, and maintaining it (and implicitly, getting it right in the first place).

Reply:
Interesting, I was trying to further simplify those types to achieve uniformity as well. But I got stuck because the details of the algorithm aren't crystal clear to me yet